plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -2546,6 +2552,62 @@ private static IcebergColumnHandle createProjectedColumnHandle(IcebergColumnHand
                 Optional.empty());
     }
 
+    @Override
+    public Optional<PartialSortApplicationResult<ConnectorTableHandle>> applyPartialSort(
+            ConnectorSession session,
+            ConnectorTableHandle tableHandle,
+            Map<ColumnHandle, io.trino.spi.connector.SortOrder> columnHandleSortOrderMap)
+    {
+        IcebergTableHandle originalTableHandle = (IcebergTableHandle) tableHandle;
+        if (originalTableHandle.isSort()) {
+            return Optional.empty();
+        }
+        IcebergTableHandle icebergTableHandle = (IcebergTableHandle) tableHandle;
+        Table table;
+        table = catalog.loadTable(session, icebergTableHandle.getSchemaTableName());
+        if (table.sortOrder().isUnsorted()) {
+            return Optional.empty();
+        }
+        SortField sortField = table.sortOrder().fields().get(0);
+        IcebergColumnHandle columnHandle = (IcebergColumnHandle) columnHandleSortOrderMap.keySet().stream().toList().get(0);
+        if (!table.schema().findColumnName(sortField.sourceId()).equalsIgnoreCase(columnHandle.getName())) {
+            return Optional.empty();
+        }
+        io.trino.spi.connector.SortOrder trinoSortOrder = columnHandleSortOrderMap.values().stream().toList().get(0);
+        io.trino.spi.connector.SortOrder icebergSortOrder = toSortOrder(sortField);
+        boolean allSorted = true;
+        TupleDomain<IcebergColumnHandle> effectivePredicate =
+                icebergTableHandle.getUnenforcedPredicate().intersect(icebergTableHandle.getEnforcedPredicate());
+        try (CloseableIterable<FileScanTask> fileScanTasks = table.newScan().filter(toIcebergExpression(effectivePredicate)).planFiles()) {
+            for (FileScanTask scanTask : fileScanTasks) {
+                DataFile file = scanTask.file();
+                if (file.sortOrderId() == 0) {
```

Comment:
This should be iceberg  bug. Even if iceberg DataFile defines a sortorder, but the file from `planFile` does not have it, I suspect this is a bug of iceberg.

Reply:
Sorting is optional. It's legitimate for files to be sorted differently, or not be sorted at all.

However, there is a problem with `table.newScan()`. We should avoid it. We use it for stats today, and we know it's causing problems (takes too long), so we don't want to do this during planning.

Reply:
File sorting is indeed optional, but necessary for this PR to accelerate queries (we can support partial file sorting scenarios later). Since `table.newScan()` is not recommended, is there another way to get information about whether the files have been sorted during the planning stage ?

And For query acceleration scenarios based on large amounts of top-N data, I think the planning time is negligible ?